cmd/ncproxy/service.go

```diff
@@ -0,0 +1,221 @@
+	// it when it panics. Remember the old stderr to restore it before removing
+	// the panic file.
+	sh := uint32(windows.STD_ERROR_HANDLE)
+	h, err := windows.GetStdHandle(sh)
+	if err != nil {
+		return err
+	}
+	oldStderr = h
+
+	if err := windows.SetStdHandle(sh, windows.Handle(panicFile.Fd())); err != nil {
+		return err
+	}
+
+	// Reset os.Stderr to the panic file (so fmt.Fprintf(os.Stderr,...) actually gets redirected)
+	os.Stderr = os.NewFile(panicFile.Fd(), "/dev/stderr")
+
+	// Force threads that panic to write to stderr (the panicFile handle now).
+	log.SetOutput(os.Stderr)
+	return nil
+}
+
+func removePanicFile() {
+	if st, err := panicFile.Stat(); err == nil {
+		if st.Size() == 0 {
+			sh := uint32(windows.STD_ERROR_HANDLE)
+			_ = windows.SetStdHandle(sh, oldStderr)
+			_ = panicFile.Close()
+			_ = os.Remove(panicFile.Name())
+		}
+	}
+}
+
+func registerService() error {
+	p, err := getServicePath()
+	if err != nil {
+		return err
+	}
+	m, err := mgr.Connect()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = m.Disconnect()
+	}()
+
+	c := mgr.Config{
+		ServiceType:  windows.SERVICE_WIN32_OWN_PROCESS,
+		StartType:    mgr.StartAutomatic,
+		ErrorControl: mgr.ErrorNormal,
+		DisplayName:  "Ncproxy",
+		Description:  "Network configuration proxy",
+	}
+
+	// Configure the service to launch with the arguments that were just passed.
+	args := []string{"-run-service"}
+	for _, a := range os.Args[1:] {
+		if a != "-register-service" {
+			args = append(args, a)
+		}
+	}
+
+	s, err := m.CreateService(serviceName, p, c, args...)
+	if err != nil {
+		return err
+	}
+	defer s.Close()
+
+	t := []scAction{
+		{Type: scActionRestart, Delay: uint32(15 * time.Second / time.Millisecond)},
+		{Type: scActionRestart, Delay: uint32(15 * time.Second / time.Millisecond)},
+		{Type: scActionNone},
+	}
+	lpInfo := serviceFailureActions{ResetPeriod: uint32(24 * time.Hour / time.Second), ActionsCount: uint32(3), Actions: uintptr(unsafe.Pointer(&t[0]))}
+	return windows.ChangeServiceConfig2(s.Handle, serviceConfigFailureActions, (*byte)(unsafe.Pointer(&lpInfo)))
+}
+
+func unregisterService() error {
+	m, err := mgr.Connect()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = m.Disconnect()
+	}()
+
+	s, err := m.OpenService(serviceName)
+	if err != nil {
+		return err
+	}
+	defer s.Close()
+
+	return s.Delete()
+}
+
+// launchService is the entry point for running ncproxy under SCM.
+func launchService(serveErr chan error) error {
+	h := &handler{
+		fromsvc:  make(chan error),
+		serveErr: serveErr,
+	}
+
+	interactive, err := svc.IsWindowsService()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		if interactive {
+			err = debug.Run(serviceName, h)
+		} else {
+			err = svc.Run(serviceName, h)
+		}
+		h.fromsvc <- err
+	}()
+
+	// Wait for the first signal from the service handler.
+	return <-h.fromsvc
+}
+
+func (h *handler) Execute(_ []string, r <-chan svc.ChangeRequest, s chan<- svc.Status) (bool, uint32) {
```

Comment:
how is this called?

Reply:
1. https://cs.opensource.google/go/x/sys/+/59db8d76:windows/svc/service.go;l=364
2. https://cs.opensource.google/go/x/sys/+/59db8d76:windows/svc/service.go;l=252